### DISCUSSION TITLE
useSearchParams is bug ?

### DISCUSSION THREAD

**Discussion by shisui0711** [vercel/next.js#69542] [Help] [answered]
### Summary

I'm not sure if it's a bug or not but in my use-case it doesn't work

### Additional information

```js
Url on browser: http://localhost:3000/recovery?token=CfDJ8KI4Kwa+FOxIlZUH5++HzLWsExBcpCGowUdV/cHL9JDpzLW+UJTctLV5D5oNjR0XyfrhKd6jI9c5Iucu6FVkzHTPB+J8j5X1aOImwKlpYyUZ5CqHjgG5hjg0/A3nlhItqmb+xII3w2pEGV404tBcZVbxmt8NnorbsAwjMYFko5siZf3UXVLwNFAJA/l6vBX1OxizKaoVo3xCxm2b/1bl3m3cTJMTcRqOuFDpAszw0n50

const token = useSearchParams().get("token");
consolog(token); //Resutl: CfDJ8KI4Kwa FOxIlZUH5  HzLWsExBcpCGowUdV/cHL9JDpzLW UJTctLV5D5oNjR0XyfrhKd6jI9c5Iucu6FVkzHTPB J8j5X1aOImwKlpYyUZ5CqHjgG5hjg0/A3nlhItqmb xII3w2pEGV404tBcZVbxmt8NnorbsAwjMYFko5siZf3UXVLwNFAJA/l6vBX1OxizKaoVo3xCxm2b/1bl3m3cTJMTcRqOuFDpAszw0n50

The plus sign has been turned into a space
```

**Comment by Juneezee** [comment]
This is not a bug on Next.js side. The standard `URL.parse` method (https://developer.mozilla.org/en-US/docs/Web/API/URL/parse_static) method also replaces `+` with space.

```javascript
const url = URL.parse("http://localhost:3000/recovery?token=CfDJ8KI4Kwa+FOxIlZUH5++HzLWsExBcpCGowUdV/cHL9JDpzLW+UJTctLV5D5oNjR0XyfrhKd6jI9c5Iucu6FVkzHTPB+J8j5X1aOImwKlpYyUZ5CqHjgG5hjg0/A3nlhItqmb+xII3w2pEGV404tBcZVbxmt8NnorbsAwjMYFko5siZf3UXVLwNFAJA/l6vBX1OxizKaoVo3xCxm2b/1bl3m3cTJMTcRqOuFDpAszw0n50");

console.log(url.searchParams.get("token"));
```

and I get the output:

```
'CfDJ8KI4Kwa FOxIlZUH5  HzLWsExBcpCGowUdV/cHL9JDpzLW UJTctLV5D5oNjR0XyfrhKd6jI9c5Iucu6FVkzHTPB J8j5X1aOImwKlpYyUZ5CqHjgG5hjg0/A3nlhItqmb xII3w2pEGV404tBcZVbxmt8NnorbsAwjMYFko5siZf3UXVLwNFAJA/l6vBX1OxizKaoVo3xCxm2b/1bl3m3cTJMTcRqOuFDpAszw0n50'
```

You need to encode your token properly, use `encodeURIComponent` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)

**Reply by Juneezee** [reply]
Further clarification here.

`+` is one of the special character that needs to be encoded, see https://developer.mozilla.org/en-US/docs/Glossary/Percent-encoding. Otherwise it is treated as a white space.

**Reply by shisui0711** [reply]
if encode and decode again then the result still does not have a + sign

I customized it like this and it worked like I wanted:
`export const useMySearchParams = () => {
  const params: Map<string,string> = new Map();
  let query = window.location.search;
  if(query.startsWith("?"))
    query = query.substring(1);
  const pairs = query.split("&");
  for(let i = 0; i < pairs.length; i++) {
    const pair = pairs[i].split("=");
    const key = decodeURIComponent(pair[0]);
    const value = decodeURIComponent(pair[1]);
    params.set(key, value);
  }
  return params;
}`

**Reply by Juneezee** [reply]
> if encode and decode again then the result still does not have a + sign

Hmm...Are you sure you are encoding and decoding correctly? I am able to get the expected result

```javascript
const raw = "CfDJ8KI4Kwa+FOxIlZUH5++HzLWsExBcpCGowUdV";
const encoded = encodeURIComponent(raw);

console.log(encoded);
// Output: "CfDJ8KI4Kwa%2BFOxIlZUH5%2B%2BHzLWsExBcpCGowUdV"

const decoded = decodeURIComponent(encoded);
console.log(decoded);
// Output: "CfDJ8KI4Kwa+FOxIlZUH5++HzLWsExBcpCGowUdV"

console.log(decoded === raw);
// Output: true
```

**Reply by shisui0711** [reply]
If you get it directly from the url it works but when you get it from useSearchParam the string has lost the + sign.
`
const expectedResult = "CfDJ8KI4Kwa+FOxIlZUH5++HzLWsExBcpCGowUdV";
const raw = useSearchParams().get("token");
console.log(raw);
//Ouput: "CfDJ8KI4Kwa FOxIlZUH5  HzLWsExBcpCGowUdV";
const encoded = encodeURIComponent(raw);

console.log(encoded);
// Output: "CfDJ8KI4Kwa%20FOxIlZUH5%20%20HzLWsExBcpCGowUdV"

const decoded = decodeURIComponent(encoded);
console.log(decoded);
// Output: "CfDJ8KI4Kwa FOxIlZUH5  HzLWsExBcpCGowUdV"

console.log(decoded === expectedResult);
// Output: false`

**Reply by Dragate** [reply, marked as answer]
@shisui0711 I think you might have misunderstood the encoding part. The token should be encoded when constructing the recovery URL, not after. `useSearchParams` seems to work as expected from my side as well, meaning it decodes correctly.

> I customized it like this and it worked like I wanted: `export const useMySearchParams = () => { const params: Map<string,string> = new Map(); let query = window.location.search; if(query.startsWith("?")) query = query.substring(1); const pairs = query.split("&"); for(let i = 0; i < pairs.length; i++) { const pair = pairs[i].split("="); const key = decodeURIComponent(pair[0]); const value = decodeURIComponent(pair[1]); params.set(key, value); } return params; }`

Apart from this being unnecessary as the issue lies elsewhere, this custom hook cannot function properly as it will error when trying to access `window` that does not exist on server-side rendering.

**Reply by shisui0711** [reply]
I got it. The url on my browser is not encoded.
I thought hooks could only be used in client components. So my custom hook should still work.

**Reply by Dragate** [reply]
You should read up on [client components](https://nextjs.org/docs/app/building-your-application/rendering/client-components#how-are-client-components-rendered). They render initially on the server and hydrate on the client. That means browser API should be handled gracefully on the server, unless you have disabled SSR for the components this hook is used in.
